src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/SemanticTokensEncoder.cs

```diff
@@ -190,8 +195,45 @@ private static void EncodeAndAddToken(ICollection<uint> encodedTokens, ITokenTex
             // Bit mask representing token modifiers applied to the current token.
             encodedTokens.Add(NoTokenModifiersMask);
 
+            // Handle control tokens only if control token list is initialized. Only for full document semantic tokens.
+            if (controlTokenDict != null)
+            {
+                HandleControlTokensData(currentToken, currentLineNumber, startPosition, startPosition + length, controlTokenDict);
+            }          
+           
             previousTokenLineNumber = currentLineNumber;
-            previousTokenStartIdx = startPosition;
+            previousTokenStartIdx = startPosition;            
+        }
+
+        /// <summary>
+        /// Encodes the token and adds it to the given <paramref name="controlTokenDict"/> dictionary.
+        /// </summary>
+        /// <param name="currentToken">Current token.</param>
+        /// <param name="currentLineNumber">current line number of the token.</param>
+        /// <param name="tokenStartPosition">Start Position of the token.</param>
+        /// <param name="tokenEndPosition">End Position of the token.</param>
+        /// <param name="controlTokenDict">Controls Token Dictionary that keeps a collection of control tokens data.</param>
+        internal static void HandleControlTokensData(ITokenTextSpan currentToken, uint currentLineNumber, uint tokenStartPosition, uint tokenEndPosition, ControlTokenDictionary controlTokenDict)
+        {
+            if (currentToken.TokenType == TokenType.Control)
```

Comment:
>if (currentToken.TokenType == TokenType.Control) [](http://example.com/codeflow?start=12&length=48)

This is the policy for adding the tokens?

So all token types of Control,  and those occur where DType.IsControl .  So the host creates these tokens by having firstName tokens of DType.IsControl =true?

Reply:
Yes. It depends on how that first name was bound during binding. Binding is the heart of these tokens. For each token, a binding is queried to decide which token type it should be decorated with. For controls, it is BindKind.Control which i think has a DType.IsControl == true.